File: packages/are-mcp-gateway/src/redaction.ts
const SECRET_PATTERNS = [
  /bearer\s+[a-z0-9._\-]+/gi,
  /sk_[a-z0-9_]{16,}/gi,
  /(api[_-]?key|token|credential|password|secret|signature)\s*[:=]\s*["']?[^"',\s]+/gi,
  /(authorization|cookie|set-cookie)\s*[:=]\s*["']?[^"',\n]+/gi
];

export function redactText(value: string): string {
  return SECRET_PATTERNS.reduce((current, pattern) => current.replace(pattern, "[REDACTED]"), value);
}

export function safeResourceLabel(input: unknown): string {
  if (input === null || input === undefined) {
    return "unknown";
  }

  if (typeof input === "string") {
    return truncate(redactText(input), 96);
  }

  if (typeof input === "number" || typeof input === "boolean") {
    return String(input);
  }

  if (Array.isArray(input)) {
    return `array/${input.length}`;
  }

  if (typeof input === "object") {
    const keys = Object.keys(input as Record<string, unknown>)
      .filter((key) => !/token|secret|password|credential|authorization|cookie|signature/i.test(key))
      .slice(0, 4)
      .join(",");
    return keys ? `object/${keys}` : "object";
  }

  return "unknown";
}

function truncate(value: string, max: number): string {
  return value.length <= max ? value : `${value.slice(0, max - 3)}...`;
}

File: python/src/are_mcp_gateway/redaction.py
from __future__ import annotations

import re
from typing import Any

SECRET_PATTERNS = [
    re.compile(r"bearer\s+[a-z0-9._\-]+", re.IGNORECASE),
    re.compile(r"sk_[a-z0-9_]{16,}", re.IGNORECASE),
    re.compile(r"(api[_-]?key|token|credential|password|secret|signature)\s*[:=]\s*[\"']?[^\"',\s]+", re.IGNORECASE),
    re.compile(r"(authorization|cookie|set-cookie)\s*[:=]\s*[\"']?[^\"',\n]+", re.IGNORECASE),
]


def redact_text(value: str) -> str:
    current = value
    for pattern in SECRET_PATTERNS:
        current = pattern.sub("[REDACTED]", current)
    return current


def safe_resource_label(value: Any) -> str:
    if value is None:
        return "unknown"
    if isinstance(value, str):
        return _truncate(redact_text(value), 96)
    if isinstance(value, bool) or isinstance(value, int) or isinstance(value, float):
        return str(value)
    if isinstance(value, list):
        return f"array/{len(value)}"
    if isinstance(value, dict):
        safe_keys = [
            key
            for key in value.keys()
            if not re.search(r"token|secret|password|credential|authorization|cookie|signature", str(key), re.IGNORECASE)
        ][:4]
        return "object/" + ",".join(map(str, safe_keys)) if safe_keys else "object"
    return "unknown"


def _truncate(value: str, limit: int) -> str:
    return value if len(value) <= limit else value[: limit - 3] + "..."
